lib/compiler-arguments.js

```diff
@@ -0,0 +1,168 @@
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+"use strict";
+
+const
+    _ = require('underscore'),
+    AWS = require('aws-sdk'),
+    logger = require('./logger').logger,
+    S3Bucket = require('./s3-handler');
+
+class CompilerArguments {
+    constructor(compilerId) {
+        this.compilerId = compilerId;
+        this.possibleArguments = [];
+        this.maxPopularArguments = 5;
+        this.storeSpecificArguments = false;
+    }
+
+    loadFromStorage(awsProps) {
+        const region = awsProps('region');
+        const bucket = awsProps('storageBucketArgStats');
+        const prefix = awsProps('storagePrefixArgStats');
+        if (region && bucket && this.compilerId) {
+            AWS.config.update({region: region});
+
+            const s3 = new S3Bucket(bucket, region);
+            s3.get(this.compilerId + ".json", prefix)
+                .then(result => {
+                    if (result.hit) {
+                        const stats = JSON.parse(result.data.toString());
+                        _.each(stats, (times, arg) => {
+                            this.addOptionToStatistics(arg, times);
+                        });
+                    } else {
+                        logger.debug(`${this.compilerId}.json not present in storage`);
+                    }
+                });
+        }
+    }
+
+    getPopularArguments(excludeUsedArguments) {
+        let possibleArguments = {};
+        if (excludeUsedArguments && excludeUsedArguments.length > 0) {
+            _.forEach(this.possibleArguments, (obj, argKey) => {
+                for (var argIdx in excludeUsedArguments) {
+                    if (this.match(argKey, excludeUsedArguments[argIdx])) return;
+                }
+                possibleArguments[argKey] = obj;
+            });
+        } else {
+            possibleArguments = this.possibleArguments;
+        }
+
+        let arr = _.pairs(possibleArguments);
+        arr.sort((a, b) => {
+            if ((a[1].timesused === 0) && (b[1].timesused === 0)) {
```

Comment:
How useful do you think the "timesused" thing is? Seems it will differ from one instance to another, and restarts lose the usage, so it would seem pretty random. I wonder if it's worth it? Or whether we should "just" move to an actual persistent database, shared across instances for this. Not a blocker to merging, but makes me wonder if it will give surprising results on the live site as requests get round-robined between instances.

Reply:
It's important for sorting, it brings the most used arguments to the top, but once that's done the JS Object behaviour will take over and put keys wherever it likes to put keys. So maybe it's not too useful when sending over the client, but I think I prefer a consistent layout from start to end, even though it's not sorted clientside.

Reply:
Ah! I absolutely agree the most used should be at the top. But that the random differences each node will see, coupled with the counts from the static analysis presumably being very very high, means the small adjustments made here for local statistics are probably unnecessarily. You're right, without "timesused" we don't get the statically determined relevance. I suppose I'm trying to say the dynamically updated "timesused" probably doesn't help. Though I guess it's the only thing one has when running locally?

Reply:
WHoops: we spoke in person and I completely misread the code. Sorry! :) This doesn't happen at all!